[Merged by Bors] - chore(algebra/associated): use more dot notation

Vierkantor opened the pull request:
I was getting annoyed that working with definitions such as `irreducible`, `prime` and `associated`, I had to write quite verbose terms like `dvd_symm_of_irreducible (irreducible_of_prime (prime_of_factor _ hp)) (irreducible_of_prime (prime_of_factor _ hq)) hdvd`, where a lot of redundancy can be eliminated with dot notation: `(prime_of_factor _ hp).irreducible.dvd_symm (prime_of_factor _ hq).irreducible hdvd`. This PR changes the spelling of many of the lemmas in `algebra/associated.lean` to make usage of dot notation easier. It also adds a few shortcut lemmas that address common patterns.

Since this change touches a lot of files, I'll add a RFC label / [open a thread on Zulip](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/RFC.3A.20adding.20dot.20notations.20.238556) to see what everyone thinks.

Renamings:
 * `irreducible_of_prime` → `prime.irreducible`
 * `dvd_symm_of_irreducible` → `irreducible.dvd_symm`
 * `dvd_symm_iff_of_irreducible` → `irreducible.dvd_comm` (cf. `eq.symm` versus `eq_comm`)
 * `associated_mul_mul` → `associated.mul_mul`
 * `associated_pow_pow` → `associated.pow_pow`
 * `dvd_of_associated` → `associated.dvd`
 * `dvd_dvd_of_associated` → `associated.dvd_dvd`
 * `dvd_iff_dvd_of_rel_{left,right}` → `associated.dvd_iff_dvd_{left,right}`
 * `{eq,ne}_zero_iff_of_associated` → `associated.{eq,ne}_zero_iff`
 * `{irreducible,prime}_of_associated` → `associated.{irreducible,prime}`
 * `{irreducible,is_unit,prime}_iff_of_associated` → `associated.{irreducible,is_unit,prime}_iff`
  * `associated_of_{irreducible,prime}_of_dvd` → `{irreducible,prime}.associated_of_dvd`
 * `gcd_eq_of_associated_{left,right}` → `associated.gcd_eq_{left,right}`
 * `{irreducible,prime}_of_degree_eq_one_of_monic` → `monic.{irreducible,prime}_of_degree_eq_one`
  * `gcd_monoid.prime_of_irreducible` → `irreducible.prime` (since the GCD case is probably the only case we care about for this implication. And we could generalize to Schreier domains if not)

Additions:
 * `associated.is_unit := (associated.is_unit_iff _).mp` (to match `associated.prime` and `associated.irreducible`)
 * `associated.mul_left := associated.mul_mul (associated.refl _) _`
 * `associated.mul_right := associated.mul_mul _  (associated.refl _)`

Other changes:
 * `associated_normalize`, `normalize_associated`, `associates.mk_normalize`, `normalize_apply`, `prime_X_sub_C`: make their final parameter explicit, since it is only inferrable when trying to apply these lemmas. This change helped to golf a few proofs from tactic mode to term mode.
 * slight golfing and style fixes

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
chore(algebra/associated): use more dot notation

eric-wieser commented on src/algebra/associated.lean:
```diff
@@ -187,17 +187,17 @@ have hpd : p ∣ x * y, from ⟨z, by rwa [mul_right_inj' hp0] at h⟩,
 end
 
-lemma dvd_symm_iff_of_irreducible [monoid α] {p q : α}
+lemma irreducible.dvd_symm_iff [monoid α] {p q : α}
```
Possibly `dvd_comm` would be better, cf eq.symm vs eq_comm

eric-wieser reviewed (approved):
I've not used this API much, but these changes all look very sensible to me. I'll let someone else sign off since this is marked RFC.

Vierkantor committed:
Lint fix: overlong line

YaelDillies commented on src/algebra/associated.lean:
```diff
@@ -163,7 +163,7 @@ begin
 end
 
-lemma irreducible_of_prime [comm_cancel_monoid_with_zero α] {p : α} (hp : prime p) :
+lemma prime.irreducible [comm_cancel_monoid_with_zero α] {p : α} (hp : prime p) :
```
```suggestion
protected lemma prime.irreducible [comm_cancel_monoid_with_zero α] {p : α} (hp : prime p) :
```
Probably necessary to avoid weird errors!

YaelDillies commented on src/field_theory/minpoly.lean:
```diff
@@ -399,7 +399,7 @@ lemma root {x : B} (hx : is_integral A x) {y : A} (h : is_root (minpoly A x) y)
 have key : minpoly A x = X - C y :=
 eq_of_monic_of_associated (monic hx) (monic_X_sub_C y) (associated_of_dvd_dvd
-  (dvd_symm_of_irreducible (irreducible_X_sub_C y) (irreducible hx) (dvd_iff_is_root.2 h))
+  ((irreducible_X_sub_C y).dvd_symm (irreducible hx) (dvd_iff_is_root.2 h))
```
```suggestion
  ((irreducible_X_sub_C y).dvd_symm hx.irreducible (dvd_iff_is_root.2 h))
```
Maybe this works/is something you could make work?

Vierkantor replied:
I believe this is `minpoly.irreducible (hx : is_integral A x) : irreducible (minpoly A x)`. I would make it protected if there is any ambiguity, but I don't think we should be changing the namespace.

YaelDillies replied:
Oh yeah, I was expecting something else.

YaelDillies commented on src/ring_theory/discrete_valuation_ring.lean:
```diff
@@ -398,7 +398,7 @@ lemma add_val_def (r : R) (u : units R) {ϖ : R} (hϖ : irreducible ϖ) (n : ℕ
 by rw [add_val, add_valuation_apply, hr,
     eq_of_associated_left (associated_of_irreducible R hϖ
-      (irreducible_of_prime (classical.some_spec (exists_prime R)))),
+      ((classical.some_spec (exists_prime R)).irreducible)),
```
```suggestion
      (classical.some_spec (exists_prime R)).irreducible),
```
Dot notation, dot notation, spare us parenthesation.

Vierkantor committed:
fix

Vierkantor committed:
`dvd_symm_iff` → `dvd_comm` (doesn't seem to be used anywhere?)

Vierkantor committed:
protect `prime.irreducible` against confusion

Vierkantor committed:
Don't need these parens

Vierkantor committed:
Fix

YaelDillies commented on src/algebra/associated.lean:
```diff
@@ -242,22 +242,30 @@ theorem associated_one_of_associated_mul_one [comm_monoid α] {a b : α} :
 | ⟨c₁, h₁⟩ ⟨c₂, h₂⟩ := ⟨c₁ * c₂, by simp [h₁.symm, h₂.symm, mul_assoc, mul_comm, mul_left_comm]⟩
 
-lemma associated_pow_pow [comm_monoid α] {a b : α} {n : ℕ} (h : a ~ᵤ b) :
+lemma associated.mul_left [comm_monoid α] (a₁ : α) {a₂ b₂ : α} (h : a₂ ~ᵤ b₂) :
```
Wouldn't those names make more sense?
```suggestion
lemma associated.mul_left [comm_monoid α] (c : α) {a b : α} (h : a ~ᵤ b) :
```
or this maybe?
```suggestion
lemma associated.mul_left [comm_monoid α] (a : α) {b c : α} (h : b ~ᵤ c) :
```
Same below

YaelDillies commented on src/algebra/associated.lean:
```diff
@@ -286,64 +294,66 @@ multiset.induction_on s (by simp [mt is_unit_iff_dvd_one.2 hp.not_unit])
     ⟨λ ⟨v, hv⟩, hp.not_unit ⟨v * u⁻¹, by simp [hv, hu.symm]⟩,
       hu ▸ by { simp [units.mul_right_dvd], intros a b, exact hp.div_or_div }⟩⟩
 
 lemma associated_of_irreducible_of_dvd [cancel_monoid_with_zero α] {p q : α}
```
Don't know if this one is worth the dot.
```suggestion
lemma irreducible.associated_of_dvd [cancel_monoid_with_zero α] {p q : α}
```
or, with a shuffling of the arguments,
```suggestion
lemma dvd.associated_of_irreducible [cancel_monoid_with_zero α] {p q : α}
```
or even, as being irreducible is a clear condition (is it?),
```suggestion
lemma dvd.associated [cancel_monoid_with_zero α] {p q : α}
```

Vierkantor replied:
I'm going for the first option, the `irreducible` namespace, since there are other cases where division implies the symmetric case as well: units, prime elements (see the next declaration!), ...

Vierkantor replied:
These don't seem to be used anywhere, by the way...

YaelDillies commented on src/algebra/associated.lean:
```diff
@@ -286,64 +294,66 @@ multiset.induction_on s (by simp [mt is_unit_iff_dvd_one.2 hp.not_unit])
-⟨irreducible_of_associated h, irreducible_of_associated h.symm⟩
+⟨h.irreducible, h.symm.irreducible⟩
 
 lemma associated_mul_left_cancel [comm_cancel_monoid_with_zero α] {a b c d : α}
```
```suggestion
lemma associated.mul_left_cancel [comm_cancel_monoid_with_zero α] {a b c d : α}
```
could be worth a shot. Do we like it when one can only dot notate on certain kinds of arguments? If we don't, then we can exchange `h` and `h₁`.

Vierkantor replied:
I'm not so convinced by this name, since it would suggest something like `associated a b → c * a = c * b → a = b`. How about:
```suggestion
lemma associated.of_mul_left [comm_cancel_monoid_with_zero α] {a b c d : α}
```

Vierkantor replied:
(And idem for `associated_mul_right_cancel` → `associated.of_mul_right` below)

YaelDillies replied:
Sure! I think `associated.mul_left_cancel` would be pretty clear (modulo a minimum of intuition about association), but this is fine too.

YaelDillies commented on src/algebra/gcd_monoid.lean:
```diff
@@ -592,14 +592,14 @@ begin
 
 theorem irreducible_iff_prime {p : α} : irreducible p ↔ prime p :=
-⟨prime_of_irreducible, irreducible_of_prime⟩
+⟨prime_of_irreducible, prime.irreducible⟩
```
Definitely change this one too!
```suggestion
⟨irreducible.prime, prime.irreducible⟩
```
(l.588 is out of the scope of the diff, so I can't suggest on the statement `prime_of_irreducible` directly)

Vierkantor replied:
That won't work with dot notation, since it's declared in the `gcd_monoid` namespace. (I don't think `open gcd_monoid` would help either.)

YaelDillies replied:
Arf, yeah :/
But does it really need to be under the `gcd_monoid` namespace? Are there other ways of getting primality out of irreducibility?

Vierkantor replied:
I don't know from the top of my head any interesting case other than `gcd_monoid` where all irreducibles are prime, but I don't see `irreducible → prime` implying the existence of a gcd either.

Vierkantor replied:
Ah, I found the condition: being a [Schreier domain](https://en.wikipedia.org/wiki/Schreier_domain)

YaelDillies replied:
Ahah, great find! Then if we ever define Schreier domains in mathlib, this one could be generalised. Could you then move it out of the `gcd_monoid` namespace and put a TODO for the generalisation?

Vierkantor replied:
Sure, I guess I don't have strong feelings on whether to keep this namespaced until we define Schreier domains (monoids?).

Vierkantor replied:
I'm actually going to revert this change, because there's a couple of name conflicts between `gcd_monoid.irreducible_iff_prime` and `unique_factorization_monoid.irreducible_iff_prime`. And UFM implies GCD-monoid only if you choose a representative for each equivalence class up to associated. So marking `irreducible_iff_prime` protected doesn't quite work.

If you define Schreier domains, feel free to revert the revert! :P

YaelDillies commented on src/data/polynomial/ring_division.lean:
```diff
@@ -144,14 +144,14 @@ have p = X - C (- p.coeff 0),
 
 theorem irreducible_X_sub_C (r : R) : irreducible (X - C r) :=
-irreducible_of_prime prime_X_sub_C
+prime.irreducible prime_X_sub_C
```
Does that not work? (untested) Same for `polynomial.irreducible_X` and `padic_int.irreducible_p`.
```suggestion
prime_X_sub_C.irreducible
```
Also, do you want to make the explicitness of `r` consistent between `polynomial.prime_X_sub_C` and `polynomial.irreducible_X_sub_C`?

Vierkantor replied:
This one works only if `r` is explicit. Otherwise, the elaborator has to make too many guesses to figure out what the `.irreducible` dot notation is applied to. So `prime_X.irreducible` below would not work.

YaelDillies commented on src/data/polynomial/ring_division.lean:
```diff
@@ -144,14 +144,14 @@ have p = X - C (- p.coeff 0),
-irreducible_of_prime prime_X
+prime.irreducible prime_X
 
 lemma irreducible_of_degree_eq_one_of_monic (hp1 : degree p = 1)
```
What about
```suggestion
lemma monic.irreducible_of_degree_eq_one (hp1 : degree p = 1)
```

Vierkantor replied:
Sure, I'll do this one (and the `prime` above too), but for more of these changes, please open a new PR!

YaelDillies commented on src/field_theory/polynomial_galois_group.lean:
```diff
@@ -332,7 +332,7 @@ begin
-  { have key := dvd_symm_of_irreducible (minpoly.irreducible hα) p_irr this,
+  { have key := (minpoly.irreducible hα).dvd_symm p_irr this,
     apply le_antisymm,
     { exact nat_degree_le_of_dvd this p_irr.ne_zero },
```
I feel like `nat_degree_le_of_dvd ` could become `dvd.nat_degree_le`

Vierkantor replied:
That sounds like a perfect change for a new PR :)

YaelDillies commented on src/ring_theory/unique_factorization_domain.lean:
```diff
@@ -182,7 +182,7 @@ by haveI := classical.dec_eq α; exact
       (irreducible_iff_prime.1 (hf p (by simp)))
       (λ q hq, irreducible_iff_prime.1 (hg _ hq)) $
-        (dvd_iff_dvd_of_rel_right hfg).1
+        (hfg.dvd_iff_dvd_right).1
```
```suggestion
        hfg.dvd_iff_dvd_right.1
```

YaelDillies commented on src/ring_theory/polynomial/content.lean:
```diff
@@ -415,7 +415,7 @@ begin
   rw [sub_add_cancel, ← rprim.dvd_prim_part_iff_dvd (mul_ne_zero hC0 s0)] at h,
   rcases is_unit_prim_part_C r.leading_coeff with ⟨u, hu⟩,
-  apply dvd.trans h (dvd_of_associated (associated.symm ⟨u, _⟩)),
+  apply dvd.trans h (associated.symm ⟨u, _⟩).dvd,
```
```suggestion
  apply h.trans (associated.symm ⟨u, _⟩).dvd,
```

YaelDillies commented on src/ring_theory/discrete_valuation_ring.lean:
```diff
@@ -444,10 +444,10 @@ begin
     rw [add_val, add_valuation_apply, add_valuation_apply,
       multiplicity_le_multiplicity_iff] at h,
-    exact dvd.trans (dvd_of_associated ha) (h n (dvd_of_associated ha.symm)), },
+    exact dvd.trans ha.dvd (h n ha.symm.dvd), },
```
```suggestion
    exact ha.dvd.trans (h n ha.symm.dvd), },
```

YaelDillies commented on src/field_theory/minpoly.lean:
```diff
@@ -290,7 +290,7 @@ theorem unique' [nontrivial B] {p : polynomial A} (hp1 : _root_.irreducible p)
 let ⟨q, hq⟩ := dvd A x hp2 in
 eq_of_monic_of_associated hp3 (monic ⟨p, ⟨hp3, hp2⟩⟩) $
-mul_one (minpoly A x) ▸ hq.symm ▸ associated_mul_mul (associated.refl _) $
+mul_one (minpoly A x) ▸ hq.symm ▸ (associated.refl _).mul_mul $
```
```suggestion
mul_one (minpoly A x) ▸ hq.symm ▸ associated.mul_left _ $
```

YaelDillies commented on src/ring_theory/unique_factorization_domain.lean:
```diff
@@ -102,7 +102,7 @@ wf_dvd_monoid.induction_on_irreducible a
     ⟨i ::ₘ s, ⟨by clear _let_match; finish,
       by { rw multiset.prod_cons,
-           exact associated_mul_mul (by refl) hs.2 }⟩⟩)
+           exact (associated.refl _).mul_mul hs.2 }⟩⟩)
```
```suggestion
           exact hs.2.mul_left _ }⟩⟩)
```

YaelDillies commented on src/ring_theory/unique_factorization_domain.lean:
```diff
@@ -335,20 +335,20 @@ theorem irreducible_iff_prime_of_exists_unique_irreducible_factors [comm_cancel_
             calc multiset.prod (p ::ₘ fx)
                   ~ᵤ a * b : by rw [hx, multiset.prod_cons];
-                    exact associated_mul_mul (by refl) hfx.2
+                    exact (associated.refl _).mul_mul hfx.2
```
```suggestion
                    exact hfx.2.mul_left _
```

YaelDillies commented on src/ring_theory/unique_factorization_domain.lean:
```diff
@@ -424,8 +424,7 @@ have multiset.rel associated (p ::ₘ factors b) (factors a),
-        by rw multiset.prod_cons; exact associated_mul_mul
-          (associated.refl _)
+        by rw multiset.prod_cons; exact (associated.refl _).mul_mul
           (associated.symm (factors_prod hb0))),
```
```suggestion
        by rw multiset.prod_cons; exact (factors_prod hb0).symm.mul_left _)
```

YaelDillies commented on src/ring_theory/unique_factorization_domain.lean:
```diff
@@ -460,7 +459,7 @@ begin
   { rw multiset.prod_add,
-    exact associated.trans (associated_mul_mul (factors_prod hx) (factors_prod hy))
+    exact associated.trans ((factors_prod hx).mul_mul (factors_prod hy))
       (factors_prod (mul_ne_zero hx hy)).symm, }
```
```suggestion
    exact ((factors_prod hx).mul_mul (factors_prod hy)).trans
      (factors_prod (mul_ne_zero hx hy)).symm }
```

YaelDillies commented on src/ring_theory/unique_factorization_domain.lean:
```diff
@@ -616,8 +615,8 @@ begin
-    apply dvd.trans (dvd_of_associated (associated_pow_pow _)) (dvd.intro u.prod rfl),
+    rw [← (factors_prod hb).dvd_iff_dvd_right, hu, prod_add, prod_repeat],
+    apply dvd.trans ((associated.pow_pow _).dvd) (dvd.intro u.prod rfl),
     apply associated_normalize }
```
```suggestion
    apply (associated.pow_pow _).dvd.trans (dvd.intro u.prod rfl),
    apply associated_normalize }
```
or maybe even
```suggestion
    exact (associated.pow_pow $ associated_normalize _).dvd.trans (dvd.intro u.prod rfl) }
```

YaelDillies reviewed (approved):
Wow this is a definitive improvement!

I've left comments in the few places where we could push even more. And then I guess, LGTM!

Vierkantor committed:
`associated_of_{irreducible,prime}_of_dvd` → `{irreducible,prime}.associated_of_dvd`

Co-Authored-By: Yaël Dillies <yael.dillies@gmail.com>

Vierkantor committed:
Make the constant term explicit in `prime_X_sub_C`

Vierkantor committed:
More dot notation

Vierkantor committed:
Shortcut!

Co-Authored-By: Yaël Dillies <yael.dillies@gmail.com>

Vierkantor committed:
Shortcut all the things!

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

Vierkantor committed:
Fix

Vierkantor committed:
`associated_mul_{left,right}_cancel` → `associated.of_mul_{left,right}`

Vierkantor commented on src/algebra/associated.lean:
```diff
@@ -242,22 +242,30 @@ theorem associated_one_of_associated_mul_one [comm_monoid α] {a b : α} :
+
+lemma associated.mul_right [comm_monoid α] {a₁ b₁ : α} (h : a₁ ~ᵤ b₁) (a₂ : α) :
+  (a₁ * a₂) ~ᵤ (b₁ * a₂) :=
+h.mul_mul (associated.refl a₂)
```
```suggestion
lemma associated.mul_left [comm_monoid α] (a : α) {b c : α} (h : b ~ᵤ c) :
  (a * b) ~ᵤ (a * c) :=
(associated.refl a).mul_mul h

lemma associated.mul_right [comm_monoid α] {a b : α} (h : a ~ᵤ b) (c : α) :
  (a * c) ~ᵤ (b * c) :=
h.mul_mul (associated.refl c)
```

Vierkantor committed:
Better variable names

Vierkantor committed:
Unnamespace this

Vierkantor committed:
Revert "Unnamespace this"

This reverts commit 8f5295f97cd0bfc6375869753517e1d8d9a68f4a.

Vierkantor committed:
Fix

YaelDillies committed:
fix

Vierkantor committed:
Dot notation for `dvd.trans`

This was a mostly automatic change (using 2 vim macros), so I wouldn't be surprised if things broke
